data-url/src/lib.rs

```diff
@@ -14,6 +14,19 @@
 //! assert_eq!(body, b"Hello World!");
 //! assert!(fragment.is_none());
 //! ```
+#![no_std]
+
+// For forwards compatibility
+#[cfg(feature = "std")]
+extern crate std as _;
```

Comment:
Could you explain how that works? Or point me to a resource that does that?

Reply:
If we didn't have this line, a user could add a dependency on `data-url` on a platform like [the Arduino](https://github.com/Rahix/avr-hal), and since we've specified `#![no_std]` in the crate, compilation will succeed, even though the `std` feature is enabled.

Let's say that in the future, we wanted to add an error type to this crate, and have it implement `std::error::Error`. To do so, we'd have to introduce a `#[cfg(feature = "std")] extern crate std;`. But since the hypothetical Arduino user didn't specify `default-features = false`, their crate would now fail to build.

With this line in place to begin with, we know the user has specified `default-features = false` to even compile the library.

For reference, the error looks something like this (can be simulated with `-Zbuild-std=core,alloc`):

```
error[E0463]: can't find crate for `std`
 --> data-url/src/lib.rs:21:1
  |
  | extern crate std as _;
  | ^^^^^^^^^^^^^^^^^^^^^^ can't find crate
  |
```

---

Did that clear it up? Or do you want me to try explaining it in a different way? Unfortunately, I haven't found many resources for "building Rust libraries that work well with `no_std`", so I can't really link to a blog post or something explaining the details, this is mostly figured out by trial and error :/